pricegraph/wasm/tests/nodejs.rs

```diff
@@ -0,0 +1,37 @@
+extern crate wasm_bindgen_test;
+
+#[path = "../../data/mod.rs"]
+mod data;
+
+use dex_pricegraph::PriceEstimator;
+use wasm_bindgen::prelude::*;
+use wasm_bindgen_test::*;
+
+#[wasm_bindgen]
+extern "C" {
+    #[wasm_bindgen(js_namespace = Date)]
+    fn now() -> f64;
+}
+
+fn time<T>(f: impl FnOnce() -> T) -> (T, f64) {
+    let start = now();
+    let result = f();
+    (result, now() - start)
+}
+
+#[wasm_bindgen_test]
+fn estimate_price() {
+    let (estimator, load_time) = time(|| PriceEstimator::new(&*data::DEFAULT_ORDERBOOK).unwrap());
+    let (price, estimate_time) = time(|| {
+        estimator
+            .estimate_price(7, 1, 100.0 * 10.0f64.powi(18))
+            .unwrap()
+    });
+
+    console_log!(
+        "DAI-WETH price for selling 100 WETH: 1 WETH = {} DAI (load {}ms, estimate {}ms)",
+        price,
+        load_time,
+        estimate_time,
+    );
```

Comment:
How is this actually run in js and what is the unwrap translated to?

Reply:
The `nodejs.rs` gets compiled to WASM and then a `node` executable is spun up that executes the WASM module's `#[wasm_bindgen_test]` methods.

In `PriceEstimator::new` there is a magic call to `console_error_panic_hook::set_once();` that handles panics by logging to `console.error`. This is pretty much the best we can do with panics since they produce an `unreachable` instruction that causes a trap the WASM runtime and makes the method return a `RuntimeError` right away (meaning you can't `catch_unwind` these panics and wrap it in a `Result`).

`Result`s in general produce JavaScript exceptions on `Err`, which is what we use for normal errors (in the `PriceEstimator::new` method for example). The hope is that the code should never panic (as that currently *should* only happen if some of the assumptions I made were wrong or there is a logic bug, and not on, say bad user input)

I must admit the amount of magic happening here is pretty crazy - gotta love `proc_macro`s and code generation!

Reply:
Here are some resource I found helpful:
- https://rustwasm.github.io/docs/book/
- https://rustwasm.github.io/docs/wasm-bindgen/

I also found running `wasm-pack build pricegraph/wasm --scope gnosis.pm` and then inspecting the `pricegraph/wasm/pkg` directory quite interesting to see how the JS/Rust interface works under the hood.